chain/chunks/src/lib.rs

```diff
@@ -921,6 +921,12 @@ impl ShardsManager {
             return Err(Error::InvalidPartMessage);
         }
 
+        // check part merkle proofs
+        let num_total_parts = self.runtime_adapter.num_total_parts();
```

Comment:
any reason for moving this block of code?

Reply:
This check doesn't require looking up the epoch id. I thought it might be good to do as many check as possible before hitting the missing block error.

Reply:
I think it should be the other way around: the part of the processing that happens before the signature validation we do for each received chunk.

So if someone is sending invalid chunks (or many valid chunks maliciously created for the same height), if the signature verification goes first, we will only check the signature for each of them. Otherwise we will do all the steps above. And in particular verifying the merkle path is not very cheap (though may be not too expensive compared to the sig verification).

Reply:
But isn't it easier to send chunks for blocks that don't exist if we put the signature check first? All the fields could contain random bytes and we would accept the message because the parent hash is not known. If the signature check is last we will at least know this is some kind of well-formed chunk.

Reply:
I see your point, someone can send a message that has invalid data, but valid chunk hash and a signature.
Then I agree, checking the merkle proofs before the chunk signature is the correct approach.